File: src/Init/Data/String/Basic.js
const encoder = new TextEncoder();
const decoder = new TextDecoder();

export function lean_string_validate_utf8(str) {
  try {
    encoder.encode(str);
    return true;
  } catch (e) {
    return false;
  }
}

export function lean_string_data(str) {
  return encoder.encode(str);
}

export function lean_string_length(str) {
  return [...str].length;
}

export function lean_string_dec_lt(s1, s2) {
  return s1 < s2;
}

export function lean_string_is_valid_pos(str, pos) {
  const bytes = encoder.encode(str);
  return pos >= 0 && pos <= bytes.length;
}

export function lean_string_utf8_extract(str, start, len) {
  const bytes = encoder.encode(str);
  return decoder.decode(bytes.slice(start, start + len));
}

export function lean_string_utf8_get_fast(str, pos) {
  const bytes = encoder.encode(str);
  return bytes[pos] || 0;
}

export function lean_string_utf8_next_fast(str, pos) {
  const bytes = encoder.encode(str);
  if (pos >= bytes.length) return { value: null, next: pos };
  const firstByte = bytes[pos];
  let len = 0;
  if (firstByte < 0x80) len = 1;
  else if (firstByte < 0xE0) len = 2;
  else if (firstByte < 0xF0) len = 3;
  else if (firstByte < 0xF8) len = 4;
  else len = 1;
  if (pos + len > bytes.length) len = bytes.length - pos;
  const char = decoder.decode(bytes.slice(pos, pos + len));
  return { value: char, next: pos + len };
}

export function lean_string_utf8_get(str, pos) {
  const bytes = encoder.encode(str);
  return bytes[pos] || 0;
}

export function lean_string_utf8_get_opt(str, pos) {
  const bytes = encoder.encode(str);
  if (pos < 0 || pos >= bytes.length) return null;
  return bytes[pos];
}

export function lean_string_utf8_get_bang(str, pos) {
  const bytes = encoder.encode(str);
  return bytes[pos] || 0;
}

export function lean_string_utf8_next(str, pos) {
  const bytes = encoder.encode(str);
  if (pos >= bytes.length) return { value: null, next: pos };
  const firstByte = bytes[pos];
  let len = 0;
  if (firstByte < 0x80) len = 1;
  else if (firstByte < 0xE0) len = 2;
  else if (firstByte < 0xF0) len = 3;
  else if (firstByte < 0xF8) len = 4;
  else len = 1;
  if (pos + len > bytes.length) len = bytes.length - pos;
  const char = decoder.decode(bytes.slice(pos, pos + len));
  return { value: char, next: pos + len };
}

export function lean_string_utf8_prev(str, pos) {
  const bytes = encoder.encode(str);
  if (pos <= 0) return { value: null, next: pos };
  let i = pos - 1;
  while (i > 0 && (bytes[i] & 0xC0) === 0x80) {
    i--;
  }
  const firstByte = bytes[i];
  let len = 0;
  if (firstByte < 0x80) len = 1;
  else if (firstByte < 0xE0) len = 2;
  else if (firstByte < 0xF0) len = 3;
  else if (firstByte < 0xF8) len = 4;
  else len = 1;
  const char = decoder.decode(bytes.slice(i, i + len));
  return { value: char, next: i };
}

export function lean_string_utf8_at_end(str, pos) {
  const bytes = encoder.encode(str);
  return pos >= bytes.length;
}

File: src/rust/gen_init_ffi/src/ffi/Init/Data/String/Basic.rs
use leanh_l1::datatypes::{LeanObject, LeanStringObject};
use leanh_l1::emitted::{lean_box::lean_box, lean_unbox::lean_unbox};
use leanh_l1::r#priv::lean_validate_utf8::lean_validate_utf8;
use leanh_l1_initializers::r#priv::lean_sarray_size::lean_sarray_size;
// Generated stub file for Lean FFI imports
// Source: src/Init/Data/String/Basic.lean

pub fn lean_string_validate_utf8(a: *mut LeanObject) -> bool {
    let mut pos: usize = 0;
    let mut i: usize = 0;
    lean_validate_utf8(lean_sarray_cptr(a), lean_sarray_size(a), &mut pos, &mut i)
}

pub unsafe fn lean_string_data(s: *mut LeanObject) -> *mut LeanObject {
    unsafe { (*(s as *mut LeanStringObject<0>)).m_data.as_mut_ptr() as *mut LeanObject }
}

pub unsafe fn lean_string_dec_lt(a: *mut LeanObject, b: *mut LeanObject) -> u8 {
    unsafe {
        let a_ref = &*(a as *mut LeanStringObject<0>);
        let b_ref = &*(b as *mut LeanStringObject<0>);
        let a_size = a_ref.m_size.saturating_sub(1);
        let b_size = b_ref.m_size.saturating_sub(1);
        (core::slice::from_raw_parts(a_ref.m_data.as_ptr(), a_size)
            < core::slice::from_raw_parts(b_ref.m_data.as_ptr(), b_size)) as u8
    }
}

pub unsafe fn lean_string_is_valid_pos(s: *mut LeanObject, pos: *mut LeanObject) -> u8 {
    let pos = unsafe { lean_unbox(pos) };
    let size = unsafe { (*(s as *mut LeanStringObject<0>)).m_size.saturating_sub(1) };
    (pos <= size) as u8
}

// moved lean_string_utf8_extract to ffi/common/lean_string_utf8_extract.rs
// original source: Init/Data/String/Basic.rs:30-45

pub unsafe fn lean_string_utf8_get_fast(s: *mut LeanObject, pos: *mut LeanObject) -> u32 {
    let pos = unsafe { lean_unbox(pos) };
    unsafe {
        (*(s as *mut LeanStringObject<0>))
            .m_data
            .as_ptr()
            .add(pos)
            .read() as u32
    }
}

pub unsafe fn lean_string_utf8_next_fast(
    _: *mut LeanObject,
    pos: *mut LeanObject,
) -> *mut LeanObject {
    unsafe { lean_box(lean_unbox(pos) + 1) }
}

// moved lean_string_utf8_get to ffi/common/lean_string_utf8_get.rs
// original source: Init/Data/String/Basic.rs:51-53

pub unsafe fn lean_string_utf8_get_opt(
    s: *mut LeanObject,
    pos: *mut LeanObject,
) -> *mut LeanObject {
    unsafe { lean_box(lean_string_utf8_get_fast(s, pos) as usize) }
}

pub unsafe fn lean_string_utf8_get_bang(s: *mut LeanObject, pos: *mut LeanObject) -> u32 {
    unsafe { lean_string_utf8_get_fast(s, pos) }
}

// moved lean_string_utf8_next to ffi/common/lean_string_utf8_next.rs
// original source: Init/Data/String/Basic.rs:65-67

pub unsafe fn lean_string_utf8_prev(_: *mut LeanObject, pos: *mut LeanObject) -> *mut LeanObject {
    unsafe { lean_box(lean_unbox(pos).saturating_sub(1)) }
}

// moved lean_string_utf8_at_end to ffi/common/lean_string_utf8_at_end.rs
// original source: Init/Data/String/Basic.rs:88-92
